### PR TITLE
Transfer content from legacy server jobs library

### PR DESCRIPTION
**Pull request by jonatanschroeder**
As the new library is TS, this also required the convertion to TS. This also modernizes queries and converts some of the callback code to async.

### CODE REVIEW

File: apps/prairielearn/src/lib/server-jobs.ts [hunk 1]
```diff
@@ -252,10 +283,209 @@ export async function createServerJob(options: CreateServerJobOptions): Promise<
...
+        const status = job.status;
+
+        const ansiUp = new AnsiUp();
+        const output = ansiUp.ansi_to_html(liveJobs[msg.job_id]?.output ?? job.output ?? '');
+
+        callback({ status, output });
+      },
+      (err) => {
+        logger.error('socket.io joinJob error selecting job_id ' + msg.job_id, err);
+      },
```
Comment (nwalters512) [inline apps/prairielearn/src/lib/server-jobs.ts:379]: We should include `Sentry.captureException` in all these error handlers.

File: apps/prairielearn/src/lib/server-jobs.ts [hunk 2]
```diff
@@ -252,10 +283,209 @@ export async function createServerJob(options: CreateServerJobOptions): Promise<
...
+  for (const row of abandonedJobs) {
+    logger.debug('Job abandoned by server, id: ' + row.id);
+    try {
+      await queryAsync(sql.update_job_on_error, {
+        job_id: row.id,
+        output: null,
+        error_message: 'Job abandoned by server',
+      });
+    } catch (err) {
+      logger.error('errorAbandonedJobs: error updating job on error', err);
```
Comment (nwalters512) [inline apps/prairielearn/src/lib/server-jobs.ts:438]: `Sentry.captureException`.

File: apps/prairielearn/src/server.js [hunk 3]
```diff
@@ -71,7 +70,11 @@ import { markAllWorkspaceHostsUnhealthy } from './lib/workspaceHost';
 import { createSessionMiddleware } from '@prairielearn/session';
 import { PostgresSessionStore } from './lib/session-store';
 import { pullAndUpdateCourse } from './lib/course';
-import { selectJobsByJobSequenceId } from './lib/server-jobs';
+import {
+  init as serverJobsInit,
+  stop as serverJobsStop,
+  selectJobsByJobSequenceId,
+} from './lib/server-jobs';
```
Comment (nwalters512) [inline apps/prairielearn/src/server.js:77]: For consistency with others, I'd rather do one of three things:

- Continue using `import * as serverJobs ...`
- Rename `init`/`stop` to `initServerJobs`/`stopServerJobs`

I'd lean towards the former.

File: apps/prairielearn/src/lib/server-jobs.ts [hunk 4]
```diff
@@ -252,10 +283,209 @@ export async function createServerJob(options: CreateServerJobOptions): Promise<
...
+  socketServer.io.on('connection', connection);
+
+  // Start a periodic task to heartbeat all live jobs. We don't use a cronjob
+  // for this because we want this to run for every host.
+  heartbeatIntervalId = setInterval(() => {
+    const jobIds = Object.keys(liveJobs);
+    if (jobIds.length === 0) return;
+
+    queryAsync(sql.update_heartbeats, { job_ids: jobIds }).catch((err) => {
+      logger.error('Error updating heartbeats for live server jobs', err);
```
[the remark this answers is not in the document]
Comment (jonatanschroeder) [reply apps/prairielearn/src/lib/server-jobs.ts:337]: Done in 3ae6dd7b79556cd8e94d24dd781a6880faa2b001, here and in every iteration where `logger.error` was called in an exception handler.

File: apps/prairielearn/src/server.js (hunk 3 above)
Comment (jonatanschroeder) [reply apps/prairielearn/src/server.js]: Done in d85a7294b749155a6306497bec9b1ac2e446a068.

**Review by nwalters512 — APPROVED** [review]

**Pull request merged**